### PR TITLE
hostname support wildcard form

### PR DESCRIPTION
**Pull request by XiaofeiCao**
fix https://github.com/Azure/azure-sdk-for-java/issues/26885

### CODE REVIEW

File: sdk/resourcemanager/azure-resourcemanager-network/src/main/java/com/azure/resourcemanager/network/implementation/ApplicationGatewayListenerImpl.java [hunk 1]
```diff
@@ -237,7 +247,15 @@ public ApplicationGatewayListenerImpl withHttps() {
 
     @Override
     public ApplicationGatewayListenerImpl withHostname(String hostname) {
-        this.innerModel().withHostname(hostname);
+        this.innerModel().withHostname(null);
+        // future work: add hostname instead of replacing
+        if (hostname == null) {
+            this.innerModel().withHostNames(null);
```
Comment (weidongxu-microsoft) [inline sdk/resourcemanager/azure-resourcemanager-network/src/main/java/com/azure/resourcemanager/network/implementation/ApplicationGatewayListenerImpl.java:253]: Are we allowed to have no hostname in listener?

Ideally if we want to clear a hostname, code should provide a `withoutHostname` in update flow.

File: sdk/resourcemanager/azure-resourcemanager-network/src/main/java/com/azure/resourcemanager/network/models/HasHostname.java [hunk 2]
```diff
@@ -20,6 +20,8 @@
         interface WithHostname<ReturnT> {
             /**
              * Specifies the hostname to reference.
+             * The hostname supports wildcard format, e.g. "*.contoso.com", but is available only for
```
Comment (weidongxu-microsoft) [inline sdk/resourcemanager/azure-resourcemanager-network/src/main/java/com/azure/resourcemanager/network/models/HasHostname.java:23]: nit, an extra line break between description and fine detail.

File: sdk/resourcemanager/azure-resourcemanager-network/src/test/java/com/azure/resourcemanager/network/ApplicationGatewayTests.java [hunk 3]
```diff
@@ -135,6 +133,72 @@ public void canCRUDApplicationGatewayWithWAF() throws Exception {
...
+            .apply();
+
+        Assertions.assertEquals(hostname2, gateway.listeners().get(listenerName).hostname());
+
+        // override hostnames by hostname
+        gateway.innerModel().httpListeners().iterator().next().withHostNames(null).withHostname(hostname1);
+        gateway.update()
+            .apply();
+
+        Assertions.assertEquals(hostname1, gateway.listeners().get(listenerName).hostname());
```
Comment (weidongxu-microsoft) [inline sdk/resourcemanager/azure-resourcemanager-network/src/test/java/com/azure/resourcemanager/network/ApplicationGatewayTests.java:198]: Do we need this test?

Comment (XiaofeiCao) [reply sdk/resourcemanager/azure-resourcemanager-network/src/test/java/com/azure/resourcemanager/network/ApplicationGatewayTests.java:198]: It's for confirmation for service response of replacing `hostnames` with `hostname`. I wanted to try the other way around but it's much more complicated..

Comment (weidongxu-microsoft) [reply sdk/resourcemanager/azure-resourcemanager-network/src/test/java/com/azure/resourcemanager/network/ApplicationGatewayTests.java:198]: You should be able to replace this case with `withHostnames` if you support it?

File: sdk/resourcemanager/azure-resourcemanager-network/src/main/java/com/azure/resourcemanager/network/implementation/ApplicationGatewayListenerImpl.java [hunk 4]
```diff
@@ -248,17 +259,23 @@ public ApplicationGatewayListenerImpl withHttps() {
     @Override
     public ApplicationGatewayListenerImpl withHostname(String hostname) {
         this.innerModel().withHostname(null);
```
Comment (weidongxu-microsoft) [inline sdk/resourcemanager/azure-resourcemanager-network/src/main/java/com/azure/resourcemanager/network/implementation/ApplicationGatewayListenerImpl.java:261]: Move this line into `if (hostname != null) {`?

File: sdk/resourcemanager/azure-resourcemanager-network/CHANGELOG.md [hunk 5]
```diff
@@ -5,6 +5,7 @@
 ### Features Added
 
 - Supported wildcard hostname in `ApplicationGateway` listener.
+- Supported `withHostnames()` and `hostnames()` in `HasHostname`
```
Comment (weidongxu-microsoft) [inline sdk/resourcemanager/azure-resourcemanager-network/CHANGELOG.md:8]:
```suggestion
- Supported `withHostnames()` and `hostnames()` in `HasHostname`.
```

File: sdk/resourcemanager/azure-resourcemanager-network/src/main/java/com/azure/resourcemanager/network/implementation/ApplicationGatewayListenerImpl.java [hunk 6]
```diff
@@ -248,17 +259,23 @@ public ApplicationGatewayListenerImpl withHttps() {
...
             List<String> hostNames = new ArrayList<>();
             hostNames.add(hostname);
             this.innerModel().withHostNames(hostNames);
         }
         return this;
     }
 
+    @Override
+    public ApplicationGatewayListenerImpl withHostnames(List<String> hostnames) {
+        this.innerModel().withHostname(null);
```
Comment (weidongxu-microsoft) [inline sdk/resourcemanager/azure-resourcemanager-network/src/main/java/com/azure/resourcemanager/network/implementation/ApplicationGatewayListenerImpl.java:272]: Move this line into `if (hostname != null) {`?

File: sdk/resourcemanager/azure-resourcemanager-network/src/main/java/com/azure/resourcemanager/network/models/HasHostname.java [hunk 7]
```diff
@@ -4,36 +4,51 @@
...
-             * {@link ApplicationGatewaySkuName#STANDARD_V2} and {@link ApplicationGatewaySkuName#WAF_V2} SKU
              *
              * @param hostname an existing frontend name on this load balancer
              * @return the next stage of the definition
              */
             ReturnT withHostname(String hostname);
+
+            /**
+             * Specifies up to 5 hostnames to reference.
+             * @param hostnames existing frontend host names on this load balancer
```
Comment (weidongxu-microsoft) [inline sdk/resourcemanager/azure-resourcemanager-network/src/main/java/com/azure/resourcemanager/network/models/HasHostname.java:41]:
```suggestion
             * @param hostnames existing frontend hostnames on this load balancer
```

Comment (weidongxu-microsoft) [reply sdk/resourcemanager/azure-resourcemanager-network/src/main/java/com/azure/resourcemanager/network/models/HasHostname.java:41]: We should use "hostname" or "hostnames" consistently.

**Review by weidongxu-microsoft — APPROVED** [review]

File: sdk/resourcemanager/azure-resourcemanager-network/src/main/java/com/azure/resourcemanager/network/implementation/ApplicationGatewayListenerImpl.java (hunk 4 above)
Comment (XiaofeiCao) [reply sdk/resourcemanager/azure-resourcemanager-network/src/main/java/com/azure/resourcemanager/network/implementation/ApplicationGatewayListenerImpl.java]: Moved, thanks...

File: sdk/resourcemanager/azure-resourcemanager-network/src/main/java/com/azure/resourcemanager/network/implementation/ApplicationGatewayListenerImpl.java (hunk 6 above)
Comment (XiaofeiCao) [reply sdk/resourcemanager/azure-resourcemanager-network/src/main/java/com/azure/resourcemanager/network/implementation/ApplicationGatewayListenerImpl.java]: Moved.

**Review by haolingdong-msft — APPROVED** [review]

**Pull request merged**